From 4c806c9fb2e58987035d572f978d2621e2045c01 Mon Sep 17 00:00:00 2001
From: Rutger van Beusekom <rutger.van.beusekom@verum.com>
Date: Wed, 6 Nov 2019 14:56:57 +0100
Subject: [PATCH 1/4] fcntl: non blocking socket kludge for mingw

---
 libguile/filesys.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/libguile/filesys.c b/libguile/filesys.c
index db8b2423a..a8e585cf3 100644
--- a/libguile/filesys.c
+++ b/libguile/filesys.c
@@ -974,6 +974,19 @@ SCM_DEFINE (scm_fcntl, "fcntl", 2, 1, 0,
   return scm_from_int (rv);
 }
 #undef FUNC_NAME
+#else
+SCM_DEFINE (scm_fcntl, "fcntl", 2, 1, 0,
+            (SCM object, SCM cmd, SCM value),
+            "wip: assuming: cmd = F_SETFL\n, value = (logior O_NONBLOCK (fcntl object F_GETFL))")
+#define FUNC_NAME s_scm_fcntl
+{
+  unsigned long non_blocking = 1;
+  object = SCM_COERCE_OUTPORT (object);
+  SOCKET fd = _get_osfhandle(SCM_FPORT_FDES(object));
+  if(SOCKET_ERROR == ioctlsocket(fd, FIONBIO, &non_blocking))
+    SCM_SYSERROR;
+  return scm_from_int(0);
+}
 #endif /* HAVE_FCNTL */
 
 SCM_DEFINE (scm_fsync, "fsync", 1, 0, 0, 
-- 
2.24.0

From d180c6c4337a50589c41fe70f8054c9dcf0c0612 Mon Sep 17 00:00:00 2001
From: Rutger van Beusekom <rutger.van.beusekom@verum.com>
Date: Mon, 11 Nov 2019 10:18:33 +0100
Subject: [PATCH 2/4] add piped-process to allow pipelines on mingw

---
 libguile/posix.c | 114 +++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 114 insertions(+)

diff --git a/libguile/posix.c b/libguile/posix.c
index b0fcad5fd..88a667549 100644
--- a/libguile/posix.c
+++ b/libguile/posix.c
@@ -1334,6 +1334,120 @@ start_child (const char *exec_file, char **exec_argv,
 #endif
 
 #ifdef HAVE_START_CHILD
+static SCM
+scm_piped_process (SCM prog, SCM args, SCM to, SCM from)
+//to is either #f or a writeable file descriptor
+//from is either #f or a readable file descriptor
+#define FUNC_NAME "piped-process"
+{
+  //p2c[0] == (car to)
+  //p2c[1] == (cdr to)
+  //c2p[0] == (car from)
+  //c2p[1] == (cdr from)
+  
+  //long mode_bits;
+  int reading, writing;
+  int c2p[2]; /* Child to parent.  */
+  int p2c[2]; /* Parent to child.  */
+  int in = -1, out = -1, err = -1;
+  int pid;
+  char *exec_file;
+  char **exec_argv;
+  SCM read_port = SCM_BOOL_F, write_port = SCM_BOOL_F;
+
+  exec_file = scm_to_locale_string (prog);
+  exec_argv = scm_i_allocate_string_pointers (scm_cons (prog, args));
+
+  //mode_bits = scm_i_mode_bits (mode);
+  reading = scm_to_bool(from); //mode_bits & SCM_RDNG;
+  writing = scm_to_bool(to); //mode_bits & SCM_WRTNG;
+
+  if (reading)
+    {
+      c2p[0] = scm_to_int(scm_car(from));
+      c2p[1] = scm_to_int(scm_cdr(from));
+      /* if (pipe (c2p)) */
+      /*   { */
+      /*     int errno_save = errno; */
+      /*     free (exec_file); */
+      /*     errno = errno_save; */
+      /*     SCM_SYSERROR; */
+      /*   } */
+      out = c2p[1];
+    }
+  
+  if (writing)
+    {
+      p2c[0] = scm_to_int(scm_car(to));
+      p2c[1] = scm_to_int(scm_cdr(to));
+      /* if (pipe (p2c)) */
+      /*   { */
+      /*     int errno_save = errno; */
+      /*     free (exec_file); */
+      /*     if (reading) */
+      /*       { */
+      /*         close (c2p[0]); */
+      /*         close (c2p[1]); */
+      /*       } */
+      /*     errno = errno_save; */
+      /*     SCM_SYSERROR; */
+      /*   } */
+      in = p2c[0];
+    }
+  
+  {
+    SCM port;
+
+    if (SCM_OPOUTFPORTP ((port = scm_current_error_port ())))
+      err = SCM_FPORT_FDES (port);
+    if (out == -1 && SCM_OPOUTFPORTP ((port = scm_current_output_port ())))
+      out = SCM_FPORT_FDES (port);
+    if (in == -1 && SCM_OPINFPORTP ((port = scm_current_input_port ())))
+      in = SCM_FPORT_FDES (port);
+  }
+
+  pid = start_child (exec_file, exec_argv, reading, c2p, writing, p2c,
+                     in, out, err);
+
+  if (pid == -1)
+    {
+      int errno_save = errno;
+      free (exec_file);
+      if (reading)
+        {
+          close (c2p[0]);
+          close (c2p[1]);
+        }
+      if (writing)
+        {
+          close (p2c[0]);
+          close (p2c[1]);
+        }
+      errno = errno_save;
+      SCM_SYSERROR;
+    }
+
+  /* There is no sense in catching errors on close().  */
+  if (reading)
+    {
+      close (c2p[1]);
+      read_port = scm_i_fdes_to_port (c2p[0], scm_mode_bits ("r0"),
+                                      sym_read_pipe,
+                                      SCM_FPORT_OPTION_NOT_SEEKABLE);
+    }
+  if (writing)
+    {
+      close (p2c[0]);
+      write_port = scm_i_fdes_to_port (p2c[1], scm_mode_bits ("w0"),
+                                       sym_write_pipe,
+                                       SCM_FPORT_OPTION_NOT_SEEKABLE);
+    }
+
+  return scm_values (scm_list_3 (read_port,
+                                 write_port,
+                                 scm_from_int (pid)));  
+}
+  
 static SCM
 scm_open_process (SCM mode, SCM prog, SCM args)
 #define FUNC_NAME "open-process"
-- 
2.24.0

From 8bf1a574ec7eecd71daaccf78c81db5af7b60bad Mon Sep 17 00:00:00 2001
From: Rutger van Beusekom <rutger.van.beusekom@verum.com>
Date: Mon, 11 Nov 2019 11:48:32 +0100
Subject: [PATCH 3/4] fixup

---
 libguile/posix.c | 24 +++++++++++++-----------
 1 file changed, 13 insertions(+), 11 deletions(-)

diff --git a/libguile/posix.c b/libguile/posix.c
index 88a667549..599999174 100644
--- a/libguile/posix.c
+++ b/libguile/posix.c
@@ -1359,8 +1359,8 @@ scm_piped_process (SCM prog, SCM args, SCM to, SCM from)
   exec_argv = scm_i_allocate_string_pointers (scm_cons (prog, args));
 
   //mode_bits = scm_i_mode_bits (mode);
-  reading = scm_to_bool(from); //mode_bits & SCM_RDNG;
-  writing = scm_to_bool(to); //mode_bits & SCM_WRTNG;
+  reading = SCM_BOOL_T;//scm_to_bool(from); //mode_bits & SCM_RDNG;
+  writing = SCM_BOOL_T;//scm_to_bool(to); //mode_bits & SCM_WRTNG;
 
   if (reading)
     {
@@ -1431,21 +1431,22 @@ scm_piped_process (SCM prog, SCM args, SCM to, SCM from)
   if (reading)
     {
       close (c2p[1]);
-      read_port = scm_i_fdes_to_port (c2p[0], scm_mode_bits ("r0"),
-                                      sym_read_pipe,
-                                      SCM_FPORT_OPTION_NOT_SEEKABLE);
+      /* read_port = scm_i_fdes_to_port (c2p[0], scm_mode_bits ("r0"), */
+      /*                                 sym_read_pipe, */
+      /*                                 SCM_FPORT_OPTION_NOT_SEEKABLE); */
     }
   if (writing)
     {
       close (p2c[0]);
-      write_port = scm_i_fdes_to_port (p2c[1], scm_mode_bits ("w0"),
-                                       sym_write_pipe,
-                                       SCM_FPORT_OPTION_NOT_SEEKABLE);
+      /* write_port = scm_i_fdes_to_port (p2c[1], scm_mode_bits ("w0"), */
+      /*                                  sym_write_pipe, */
+      /*                                  SCM_FPORT_OPTION_NOT_SEEKABLE); */
     }
 
-  return scm_values (scm_list_3 (read_port,
-                                 write_port,
-                                 scm_from_int (pid)));  
+  return scm_from_int(pid);
+  /* return scm_values (scm_list_3 (read_port, */
+  /*                                write_port, */
+  /*                                scm_from_int (pid)));   */
 }
   
 static SCM
@@ -2451,6 +2452,7 @@ SCM_DEFINE (scm_gethostname, "gethostname", 0, 0, 0,
 static void
 scm_init_popen (void)
 {
+  scm_c_define_gsubr ("piped-process", 4, 0, 0, scm_piped_process);
   scm_c_define_gsubr ("open-process", 2, 0, 1, scm_open_process);
 }
 #endif /* HAVE_START_CHILD */
-- 
2.24.0

From 40110f6dcb3ee56c6c8f6b9bd940ff1109b37f17 Mon Sep 17 00:00:00 2001
From: Rutger van Beusekom <rutger.van.beusekom@verum.com>
Date: Tue, 12 Nov 2019 12:30:58 +0100
Subject: [PATCH 4/4] fixup

---
 module/ice-9/popen.scm | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/module/ice-9/popen.scm b/module/ice-9/popen.scm
index b166e9d0f..0f3b273cc 100644
--- a/module/ice-9/popen.scm
+++ b/module/ice-9/popen.scm
@@ -22,7 +22,7 @@
   :use-module (ice-9 threads)
   :use-module (srfi srfi-9)
   :export (port/pid-table open-pipe* open-pipe close-pipe open-input-pipe
-	   open-output-pipe open-input-output-pipe))
+	   open-output-pipe open-input-output-pipe piped-process))
 
 (eval-when (expand load eval)
   (load-extension (string-append "libguile-" (effective-version))
-- 
2.24.0

