From e3fd70e2a619ee99e1685354938f2bfcc37e2469 Mon Sep 17 00:00:00 2001
From: Johri van Eerd <johri.van.eerd@verum.com>
Date: Thu, 14 Jun 2018 11:43:14 +0200
Subject: [PATCH] Remove git magic from MCRL2Version.cmake

---
 build/cmake/MCRL2Version.cmake | 46 +++++++++++++++++++++---------------------
 1 file changed, 23 insertions(+), 23 deletions(-)

diff --git a/build/cmake/MCRL2Version.cmake b/build/cmake/MCRL2Version.cmake
index 17168f6e2..ee70475df 100644
--- a/build/cmake/MCRL2Version.cmake
+++ b/build/cmake/MCRL2Version.cmake
@@ -33,36 +33,36 @@ mark_as_advanced(MCRL2_PACKAGE_RELEASE)
 # Find Git information: git short-hash and modified/not modified
 set(MCRL2_MINOR_VERSION "Unknown")
 
-find_package(Git REQUIRED)
+#find_package(Git REQUIRED)
 # Prints the commit hash of the last commit and its short version.
-execute_process(
-  COMMAND ${GIT_EXECUTABLE} rev-list --max-count=1 --format=%h --abbrev=10 HEAD
-  WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
-  OUTPUT_VARIABLE GIT_REV_LIST
-  OUTPUT_STRIP_TRAILING_WHITESPACE
-)
+# execute_process(
+#   COMMAND ${GIT_EXECUTABLE} rev-list --max-count=1 --format=%h --abbrev=10 HEAD
+#   WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+#   OUTPUT_VARIABLE GIT_REV_LIST
+#   OUTPUT_STRIP_TRAILING_WHITESPACE
+# )
 # Convert multiple lines into a list
-STRING(REPLACE "\n" ";" GIT_REV_LIST ${GIT_REV_LIST})
+#STRING(REPLACE "\n" ";" GIT_REV_LIST ${GIT_REV_LIST})
 # The second line is the short-hash, specified by %h
-LIST(GET GIT_REV_LIST 1 GIT_COMMIT_HASH)
+#LIST(GET GIT_REV_LIST 1 GIT_COMMIT_HASH)
 
 # List the changed files compared to the last pushed commit.
-execute_process(
-  COMMAND ${GIT_EXECUTABLE} status --porcelain
-  WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
-  OUTPUT_VARIABLE GIT_FILES_CHANGED
-  OUTPUT_STRIP_TRAILING_WHITESPACE
-)
+# execute_process(
+#   COMMAND ${GIT_EXECUTABLE} status --porcelain
+#   WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+#   OUTPUT_VARIABLE GIT_FILES_CHANGED
+#   OUTPUT_STRIP_TRAILING_WHITESPACE
+# )
 
 # When the files changed are empty it was not modified.
-if(GIT_FILES_CHANGED STREQUAL "")
-  set(MCRL2_MINOR_VERSION "${GIT_COMMIT_HASH}")
-else()
-  set(MCRL2_MINOR_VERSION "${GIT_COMMIT_HASH}M")
-  if (${MCRL2_PACKAGE_RELEASE})
-    message(FATAL_ERROR "You are trying to package a release from a Git repository that has local modifications.")
-  endif()
-endif()
+# if(GIT_FILES_CHANGED STREQUAL "")
+#   set(MCRL2_MINOR_VERSION "${GIT_COMMIT_HASH}")
+# else()
+#   set(MCRL2_MINOR_VERSION "${GIT_COMMIT_HASH}M")
+#   if (${MCRL2_PACKAGE_RELEASE})
+#     message(FATAL_ERROR "You are trying to package a release from a Git repository that has local modifications.")
+#   endif()
+# endif()
 
 # Try to read build/SourceVersion, and set that version
 if(MCRL2_MINOR_VERSION STREQUAL "Unknown" AND EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/build/SourceVersion")
-- 
2.13.0

