From 37716d4cc6aa4f9482701d54afb6c6e8dc590687 Mon Sep 17 00:00:00 2001
From: Jan Nieuwenhuizen <janneke@gnu.org>
Date: Thu, 22 Feb 2018 11:51:10 +0100
Subject: [PATCH 1/5] guix: Build from git.

---
 guix.scm | 104 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 104 insertions(+)
 create mode 100644 guix.scm

diff --git a/guix.scm b/guix.scm
new file mode 100644
index 0000000..5b569fe
--- /dev/null
+++ b/guix.scm
@@ -0,0 +1,104 @@
+;;; guix.scm -- Guix package definition
+
+;;; Borrowing code from:
+;;; guile-sdl2 --- FFI bindings for SDL2
+;;; Copyright © 2015 David Thompson <davet@gnu.org>
+
+;;; Commentary:
+;;
+;; GNU Guix development package.  To build and install, run:
+;;
+;;   guix package -f guix.scm
+;;
+;; To build it, but not install it, run:
+;;
+;;   guix build -f guix.scm
+;;
+;; To use as the basis for a development environment, run:
+;;
+;;   guix environment -l guix.scm
+;;
+;;; Code:
+
+(use-modules (srfi srfi-1)
+             (srfi srfi-26)
+             (ice-9 match)
+             (ice-9 popen)
+             (ice-9 rdelim)
+             (gnu packages)
+             (gnu packages base)
+             (gnu packages ccache)
+             (gnu packages commencement)
+             (gnu packages maths)
+             (gnu packages databases)
+             (gnu packages gcc)
+             (gnu packages guile)
+             (gnu packages package-management)
+             (gnu packages version-control)
+             ((guix build utils) #:select (with-directory-excursion))
+             (guix build-system gnu)
+             (guix build-system trivial)
+             (guix gexp)
+             (guix git-download)
+             (guix svn-download)
+             ((guix licenses) #:prefix license:)
+             (guix packages))
+
+(define %source-dir (dirname (current-filename)))
+
+(define git-file?
+  (let* ((pipe (with-directory-excursion %source-dir
+                 (open-pipe* OPEN_READ "git" "ls-files")))
+         (files (let loop ((lines '()))
+                  (match (read-line pipe)
+                    ((? eof-object?)
+                     (reverse lines))
+                    (line
+                     (loop (cons line lines))))))
+         (status (close-pipe pipe)))
+    (lambda (file stat)
+      (match (stat:type stat)
+        ('directory #t)
+        ((or 'regular 'symlink)
+         (any (cut string-suffix? <> file) files))
+        (_ #f)))))
+
+(define-public mcrl2-svn
+  (let ((svn-revision 15176)
+        (version "201707")
+        (revision "1"))
+    (package
+      (inherit mcrl2)
+      (version (string-append version "." revision "." (number->string svn-revision)))
+      (source (origin
+                (method svn-fetch)
+                (uri (svn-reference
+                      (url "https://svn.win.tue.nl/repos/MCRL2/trunk")
+                      (revision svn-revision)))
+                (sha256
+                 (base32
+                  "0lizqlf4fyxxwkzp2jyva0ib63hg0dgidlxx6cvnz41fjzcf9b14"))))
+    (arguments
+     `(#:phases
+       (modify-phases %standard-phases
+         (add-after 'unpack 'set-version
+           (lambda* _
+             (substitute* '("build/cmake/MCRL2Version.cmake")
+               (("(set\\(MCRL2_MINOR_VERSION \")Unknown(\"\\))" all left right)
+                (string-append left ,(number->string svn-revision) right)))))))))))
+
+(define-public makreel.git
+  (let ((version "201707")
+        (revision "1")
+        (commit (read-string (open-pipe "git show HEAD | head -1 | cut -d ' ' -f 2" OPEN_READ))))
+    (package
+      (inherit mcrl2-svn)
+      (version (string-append version "-" revision "." (string-take commit 7)))
+      (source (local-file %source-dir #:recursive? #t #:select? git-file?))
+      (native-inputs `(("ccache" ,ccache)
+                       ("git" ,git)
+                       ("git" ,git "svn")
+                       ("subversion" ,subversion))))))
+
+;; Return it here so `guix build/environment/package' can consume it directly.
+makreel.git
-- 
Jan Nieuwenhuizen <janneke@gnu.org> | GNU LilyPond http://lilypond.org
Freelance IT http://JoyofSource.com | Avatar® http://AvatarAcademy.com

From 5e01e9e0ddd8f9d896fe776db11a57c8a18902d3 Mon Sep 17 00:00:00 2001
From: Jan Nieuwenhuizen <janneke@gnu.org>
Date: Thu, 22 Feb 2018 11:59:17 +0100
Subject: [PATCH 2/5] build: set version for building from git.

---
 build/cmake/MCRL2Version.cmake | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/build/cmake/MCRL2Version.cmake b/build/cmake/MCRL2Version.cmake
index fe8ea21..f712b67 100644
-- 
Jan Nieuwenhuizen <janneke@gnu.org> | GNU LilyPond http://lilypond.org
Freelance IT http://JoyofSource.com | Avatar® http://AvatarAcademy.com

From e0790005692b7724af09cd51494b347e6c10c013 Mon Sep 17 00:00:00 2001
From: Jan Nieuwenhuizen <janneke@gnu.org>
Date: Thu, 22 Feb 2018 14:40:16 +0100
Subject: [PATCH 3/5] tracepp: support reading traces from stdin.

* tools/release/tracepp/tracepp.cpp (input_output_tool): Hand
  istrstream to trace.load; fixes trace detection from stdin.
---
 tools/release/tracepp/tracepp.cpp | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/tools/release/tracepp/tracepp.cpp b/tools/release/tracepp/tracepp.cpp
index 52cd622..0e5c03f 100644
--- a/tools/release/tracepp/tracepp.cpp
+++ b/tools/release/tracepp/tracepp.cpp
@@ -1,6 +1,7 @@
 // Author(s): Muck van Weerdenburg
 // Copyright: see the accompanying file COPYING or copy at
 // https://github.com/mCRL2org/mCRL2/blob/master/COPYING
+// Copyright © 2018 Jan Nieuwenhuizen <janneke@gnu.org>
 //
 // Distributed under the Boost Software License, Version 1.0.
 // (See accompanying file LICENSE_1_0.txt or copy at
@@ -12,6 +13,8 @@
 #define AUTHOR "Muck van Weerdenburg"
 
 #include <iostream>
+#include <iomanip>
+#include <sstream>
 #include <fstream>
 #include <cassert>
 #include "mcrl2/trace/trace.h"
@@ -228,7 +231,9 @@ class tracepp_tool: public input_output_tool
       {
         mCRL2log(verbose) << "reading input from stdin..." << std::endl;
 
-        trace.load(std::cin);
+        std::cin >> std::noskipws;
+        std::istringstream str(std::string(std::istream_iterator<char>(std::cin), std::istream_iterator<char>()));
+        trace.load(str);
       }
       else
       {
-- 
Jan Nieuwenhuizen <janneke@gnu.org> | GNU LilyPond http://lilypond.org
Freelance IT http://JoyofSource.com | Avatar® http://AvatarAcademy.com

From cb2869e56cea991969118863a2ad1ff1d8da5ee1 Mon Sep 17 00:00:00 2001
From: Jan Nieuwenhuizen <janneke@gnu.org>
Date: Thu, 22 Feb 2018 14:47:35 +0100
Subject: [PATCH 4/5] lps2lts: support --out-action, --out-deadlock,
 --out-divergence, --out-nondeterminism.

* libraries/lts/include/mcrl2/lts/detail/lts_generation_options.h (lts_generation_options):
  Add out_action, out_deadlock, out_divergence, out_nondeterminism.
* libraries/lts/source/exploration.cpp (check_divergence): Save trace
  to m_options.out_divergence if set.
  (save_actions): Save trace to m_options.out_action if set.
  (save_nondeterministic_state): Save trace to
  m_options.out_nondeterminism if set.
  (save_deadlock): Save trace to m_options.out_deadlock if set.
* libraries/trace/include/mcrl2/trace/trace.h (save): If filename is
  `-', write to stdout.
* tools/release/lps2lts/lps2lts.cpp (parse_options): Set
  m_options.out_action, m_options.out_deadlock,
  m_options.out_divergence, m_options.out_nondeterminism from
  command-lin.
---
 .../mcrl2/lts/detail/lts_generation_options.h      |  5 ++++
 libraries/lts/source/exploration.cpp               | 17 +++++++++----
 libraries/lts/source/liblts_lts.cpp                |  2 +-
 libraries/trace/include/mcrl2/trace/trace.h        | 28 ++++++++--------------
 tools/release/lps2lts/lps2lts.cpp                  | 15 +++++++++++-
 5 files changed, 43 insertions(+), 24 deletions(-)

diff --git a/libraries/lts/include/mcrl2/lts/detail/lts_generation_options.h b/libraries/lts/include/mcrl2/lts/detail/lts_generation_options.h
index 0d1b340..396eb8f 100644
--- a/libraries/lts/include/mcrl2/lts/detail/lts_generation_options.h
+++ b/libraries/lts/include/mcrl2/lts/detail/lts_generation_options.h
@@ -1,6 +1,7 @@
 // Author(s): Jeroen Keiren
 // Copyright: see the accompanying file COPYING or copy at
 // https://github.com/mCRL2org/mCRL2/blob/master/COPYING
+// Copyright © 2018 Jan Nieuwenhuizen <janneke@gnu.org>
 //
 // Distributed under the Boost Software License, Version 1.0.
 // (See accompanying file LICENSE_1_0.txt or copy at
@@ -60,6 +61,10 @@ class lts_generation_options
     bool detect_nondeterminism;
     bool detect_divergence;
     bool detect_action;
+    std::string out_action;
+    std::string out_deadlock;
+    std::string out_divergence;
+    std::string out_nondeterminism;
     std::set < mcrl2::core::identifier_string > trace_actions;
     std::set < std::string > trace_multiaction_strings;
     std::set < mcrl2::lps::multi_action > trace_multiactions;
diff --git a/libraries/lts/source/exploration.cpp b/libraries/lts/source/exploration.cpp
index 485ab56..c996f11 100644
--- a/libraries/lts/source/exploration.cpp
+++ b/libraries/lts/source/exploration.cpp
@@ -1,6 +1,7 @@
 // Author(s): Ruud Koolen
 // Copyright: see the accompanying file COPYING or copy at
 // https://github.com/mCRL2org/mCRL2/blob/master/COPYING
+// Copyright © 2018 Jan Nieuwenhuizen <janneke@gnu.org>
 //
 // Distributed under the Boost Software License, Version 1.0.
 // (See accompanying file LICENSE_1_0.txt or copy at
@@ -735,7 +736,9 @@ void lps2lts_algorithm::check_divergence(
   {
     if (m_options.trace && m_traces_saved < m_options.max_traces)
     {
-      std::string filename = m_options.trace_prefix + "_divergence_" + std::to_string(m_traces_saved) + ".trc";
+      std::string filename = m_options.out_divergence != ""
+        ? m_options.out_divergence
+        : m_options.trace_prefix + "_divergence_" + std::to_string(m_traces_saved) + ".trc";
       if (save_trace(state_pair.state(), filename))
       {
         mCRL2log(info) << "Trace to the divergencing state is saved to '" << filename << std::endl;
@@ -777,7 +780,9 @@ void lps2lts_algorithm::save_actions(const lps::state& state, const next_state_g
         filename = filename + "_" + core::pp(a.label().name());
       }
     }
-    filename = filename + ".trc";
+    filename = m_options.out_action != ""
+      ? m_options.out_action
+      : filename + ".trc";
     if (save_trace(state, transition, filename))
       mCRL2log(info) << " and saved to '" << filename << "'";
     else
@@ -792,7 +797,9 @@ void lps2lts_algorithm::save_nondeterministic_state(const lps::state& state,
   std::size_t state_number = m_state_numbers.index(state);
   if (m_options.trace && m_traces_saved < m_options.max_traces)
   {
-    std::string filename = m_options.trace_prefix + "_nondeterministic_" + std::to_string(m_traces_saved) + ".trc";
+    std::string filename = m_options.out_nondeterminism != ""
+      ? m_options.out_nondeterminism
+      : m_options.trace_prefix + "_nondeterministic_" + std::to_string(m_traces_saved) + ".trc";
     if (save_trace(state, nondeterminist_transition, filename))
     {
       mCRL2log(info) << "Nondeterministic state found and saved to '" << filename
@@ -815,7 +822,9 @@ void lps2lts_algorithm::save_deadlock(const lps::state& state)
   std::size_t state_number = m_state_numbers.index(state);
   if (m_options.trace && m_traces_saved < m_options.max_traces)
   {
-    std::string filename = m_options.trace_prefix + "_dlk_" + std::to_string(m_traces_saved) + ".trc";
+    std::string filename = m_options.out_deadlock != ""
+      ? m_options.out_deadlock
+      : m_options.trace_prefix + "_dlk_" + std::to_string(m_traces_saved) + ".trc";
     if (save_trace(state, filename))
     {
       mCRL2log(info) << "deadlock-detect: deadlock found and saved to '" << filename
diff --git a/libraries/lts/source/liblts_lts.cpp b/libraries/lts/source/liblts_lts.cpp
index 5a4c8ab..71a0385 100644
--- a/libraries/lts/source/liblts_lts.cpp
+++ b/libraries/lts/source/liblts_lts.cpp
@@ -467,7 +467,7 @@ static void write_to_lts(const probabilistic_lts_lts_t& l, const std::string& fi
                                       action_label_list);
   t0.remove_indices();
 
-  if (filename=="")
+  if (filename=="" || filename=="-")
   {
     atermpp::write_term_to_binary_stream(t0, std::cout);
   }
diff --git a/libraries/trace/include/mcrl2/trace/trace.h b/libraries/trace/include/mcrl2/trace/trace.h
index 81a83e2..c3124a2 100644
--- a/libraries/trace/include/mcrl2/trace/trace.h
+++ b/libraries/trace/include/mcrl2/trace/trace.h
@@ -1,6 +1,7 @@
 // Author(s): Muck van Weerdenburg. Documentation Jan Friso Groote.
 // Copyright: see the accompanying file COPYING or copy at
 // https://github.com/mCRL2org/mCRL2/blob/master/COPYING
+// Copyright © 2018 Jan Nieuwenhuizen <janneke@gnu.org>
 //
 // Distributed under the Boost Software License, Version 1.0.
 // (See accompanying file LICENSE_1_0.txt or copy at
@@ -524,24 +525,15 @@ class Trace
 
     void save(std::string const& filename, TraceFormat tf = tfMcrl2)
     {
-      using std::ofstream;
-      ofstream os(filename.c_str(),ofstream::binary|ofstream::out|ofstream::trunc);
-      if (!os.is_open())
-      {
-        throw runtime_error("error saving trace (could not open file)");
-      }
-
-      try
-      {
-        save(os, tf);
-      }
-      catch (...)
-      {
-        os.close();
-        throw;
-      }
-
-      os.close();
+      if (filename == "-")
+        save(std::cout, tf);
+      else
+        {
+          std::ofstream os(filename.c_str(),std::ofstream::binary|std::ofstream::out|std::ofstream::trunc);
+          if (!os)
+            throw runtime_error("error saving trace (could not open file)");
+          save(os, tf);
+        }
     }
 
   private:
diff --git a/tools/release/lps2lts/lps2lts.cpp b/tools/release/lps2lts/lps2lts.cpp
index 0054a5c..e44ce6a 100644
--- a/tools/release/lps2lts/lps2lts.cpp
+++ b/tools/release/lps2lts/lps2lts.cpp
@@ -1,6 +1,7 @@
 // Author(s): Muck van Weerdenburg
 // Copyright: see the accompanying file COPYING or copy at
 // https://github.com/mCRL2org/mCRL2/blob/master/COPYING
+// Copyright © 2018 Jan Nieuwenhuizen <janneke@gnu.org>
 //
 // Distributed under the Boost Software License, Version 1.0.
 // (See accompanying file LICENSE_1_0.txt or copy at
@@ -80,7 +81,7 @@ class lps2lts_tool : public lps2lts_base
                    "generate an LTS from an LPS",
                    "Generate an LTS from the LPS in INFILE and save the result to OUTFILE. "
                    "If INFILE is not supplied, stdin is used. "
-                   "If OUTFILE is not supplied, the LTS is not stored.\n"
+                   "If OUTFILE is not supplied, the LTS is not stored (use - for stdout).\n"
                    "\n"
                    "If the 'jittyc' rewriter is used, then the MCRL2_COMPILEREWRITER environment "
                    "variable (default value: 'mcrl2compilerewriter') determines the script that "
@@ -172,19 +173,27 @@ class lps2lts_tool : public lps2lts_base
                  "level, and for depth first search, where NUM is the maximum depth. ").
       add_option("nondeterminism",
                  "detect nondeterministic states, i.e. states with outgoing transitions with the same label to different states. ", 'n').
+      add_option("out-nondeterminism", make_mandatory_argument("FILE-NAME"),
+                 "save nondeterminism trace to FILE-NAME. ").
       add_option("deadlock",
                  "detect deadlocks (i.e. for every deadlock a message is printed). ", 'D').
+      add_option("out-deadlock", make_mandatory_argument("FILE-NAME"),
+                 "save deadlock trace to FILE-NAME. ").
       add_option("divergence",
                  "detect divergences (i.e. for every state with a divergence (=tau loop) a message is printed). "
                  "The algorithm to detect the divergences is linear for every state, "
                  "so state space exploration becomes quadratic with this option on, causing a state "
                  "space exploration to become slow when this option is enabled. ", 'F').
+      add_option("out-divergence", make_mandatory_argument("FILE-NAME"),
+                 "save divergence trace to FILE-NAME. ").
       add_option("action", make_mandatory_argument("NAMES"),
                  "report whether an action from NAMES occurs in the transitions system, "
                  "where NAMES is a comma-separated list. A message "
                  "is printed for every occurrence of one of these action names. "
                  "With the -t flag traces towards these actions are generated. "
                  "When using -tN only N traces are generated after which the generation of the state space stops. ", 'a').
+      add_option("out-action", make_mandatory_argument("FILE-NAME"),
+                 "save action trace to FILE-NAME. ").
       add_option("multiaction", make_mandatory_argument("NAMES"),
                  "detect and report multiactions in the transitions system "
                  "from NAMES, a comma-separated list. Works like -a, except that multi-actions "
@@ -343,6 +352,10 @@ class lps2lts_tool : public lps2lts_base
       {
         parser.error("Option --suppress requires --verbose (of -v).");
       }
+      m_options.out_action = parser.option_argument("out-action");
+      m_options.out_deadlock = parser.option_argument("out-deadlock");
+      m_options.out_divergence = parser.option_argument("out-divergence");
+      m_options.out_nondeterminism = parser.option_argument("out-nondeterminism");
 
       if (2 < parser.arguments.size())
       {
-- 
Jan Nieuwenhuizen <janneke@gnu.org> | GNU LilyPond http://lilypond.org
Freelance IT http://JoyofSource.com | Avatar® http://AvatarAcademy.com

From 11be7340a85c37f216a7fc69ade48b8e4dd6553d Mon Sep 17 00:00:00 2001
From: Jan Nieuwenhuizen <janneke@gnu.org>
Date: Mon, 26 Feb 2018 10:47:58 +0100
Subject: [PATCH 5/5] ltscompare: Add option --out-counter-example=FILE-NAME.

* libraries/lts/include/mcrl2/lts/lts_algorithm.h (compare): Add
  counter_example_file_name parameter.  Update callers.
* (destructive_compare): Add counter_example_file_name parameter, use
  it to override hard coded counter example files.
---
 libraries/lts/include/mcrl2/lts/lts_algorithm.h | 30 +++++++++++++------------
 tools/release/ltscompare/ltscompare.cpp         |  8 +++++--
 2 files changed, 22 insertions(+), 16 deletions(-)

diff --git a/libraries/lts/include/mcrl2/lts/lts_algorithm.h b/libraries/lts/include/mcrl2/lts/lts_algorithm.h
index 7e54259..a1f40ee 100644
--- a/libraries/lts/include/mcrl2/lts/lts_algorithm.h
+++ b/libraries/lts/include/mcrl2/lts/lts_algorithm.h
@@ -85,7 +85,8 @@ template <class LTS_TYPE>
 bool destructive_compare(LTS_TYPE& l1,
                          LTS_TYPE& l2,
                          const lts_equivalence eq,
-                         const bool generate_counter_examples = false)
+                         const bool generate_counter_examples = false,
+                         std::string counter_example_file_name = "")
 {
   // Merge this LTS and l and store the result in this LTS.
   // In the resulting LTS, the initial state i of l will have the
@@ -233,7 +234,8 @@ template <class LTS_TYPE>
 bool compare(const LTS_TYPE& l1,
              const LTS_TYPE& l2,
              const lts_equivalence eq,
-             const bool generate_counter_examples = false);
+             const bool generate_counter_examples = false,
+             std::string counter_example_file_name = "");
 
 /** \brief Checks whether this LTS is smaller than another LTS according
  * to a preorder.
@@ -717,7 +719,7 @@ void reduce(LTS_TYPE& l,lts_equivalence eq)
 }
 
 template <class LTS_TYPE>
-bool compare(const LTS_TYPE& l1, const LTS_TYPE& l2, const lts_equivalence eq, const bool generate_counter_examples)
+bool compare(const LTS_TYPE& l1, const LTS_TYPE& l2, const lts_equivalence eq, const bool generate_counter_examples, std::string counter_example_file_name)
 {
   switch (eq)
   {
@@ -726,21 +728,21 @@ bool compare(const LTS_TYPE& l1, const LTS_TYPE& l2, const lts_equivalence eq, c
     default:
       LTS_TYPE l1_copy(l1);
       LTS_TYPE l2_copy(l2);
-      return destructive_compare(l1_copy,l2_copy,eq,generate_counter_examples);
+      return destructive_compare(l1_copy,l2_copy,eq,generate_counter_examples,counter_example_file_name);
   }
   return false;
 }
 
 template <class LTS_TYPE>
-bool compare(const LTS_TYPE& l1, const LTS_TYPE& l2, const lts_preorder pre, const bool generate_counter_example)
+bool compare(const LTS_TYPE& l1, const LTS_TYPE& l2, const lts_preorder pre, const bool generate_counter_example,std::string counter_example_file_name)
 {
   LTS_TYPE l1_copy(l1);
   LTS_TYPE l2_copy(l2);
-  return destructive_compare(l1_copy,l2_copy,pre,generate_counter_example);
+  return destructive_compare(l1_copy,l2_copy,pre,generate_counter_example,counter_example_file_name);
 }
 
 template <class LTS_TYPE>
-bool destructive_compare(LTS_TYPE& l1, LTS_TYPE& l2, const lts_preorder pre, const bool generate_counter_example)
+bool destructive_compare(LTS_TYPE& l1, LTS_TYPE& l2, const lts_preorder pre, const bool generate_counter_example,std::string counter_example_file_name)
 {
   switch (pre)
   {
@@ -799,7 +801,7 @@ bool destructive_compare(LTS_TYPE& l1, LTS_TYPE& l2, const lts_preorder pre, con
       detail::bisimulation_reduce(l2,false);
 
       // Trace preorder now corresponds to simulation preorder
-      return destructive_compare(l1,l2,lts_pre_sim,generate_counter_example);
+      return destructive_compare(l1,l2,lts_pre_sim,generate_counter_example,counter_example_file_name);
     }
     case lts_pre_weak_trace:
     {
@@ -812,13 +814,13 @@ bool destructive_compare(LTS_TYPE& l1, LTS_TYPE& l2, const lts_preorder pre, con
       detail::tau_star_reduce(l2);
 
       // Weak trace preorder now corresponds to strong trace preorder
-      return destructive_compare(l1,l2,lts_pre_trace,generate_counter_example);
+      return destructive_compare(l1,l2,lts_pre_trace,generate_counter_example,counter_example_file_name);
     }
     case lts_pre_trace_anti_chain:
     {
       if (generate_counter_example)
       {
-        detail::counter_example_constructor cec("counter_example_trace_preorder.trc");
+        detail::counter_example_constructor cec(counter_example_file_name != "" ? counter_example_file_name : "counter_example_trace_preorder.trc");
         return destructive_refinement_checker(l1, l2, trace, false, cec);
       }
       return destructive_refinement_checker(l1, l2, trace, false);
@@ -827,7 +829,7 @@ bool destructive_compare(LTS_TYPE& l1, LTS_TYPE& l2, const lts_preorder pre, con
     {
       if (generate_counter_example)
       {
-        detail::counter_example_constructor cec("counter_example_weak_trace_preorder.trc");
+        detail::counter_example_constructor cec(counter_example_file_name != "" ? counter_example_file_name : "counter_example_weak_trace_preorder.trc");
         return destructive_refinement_checker(l1, l2, trace, true, cec);
       }
       return destructive_refinement_checker(l1, l2, trace, true);
@@ -836,7 +838,7 @@ bool destructive_compare(LTS_TYPE& l1, LTS_TYPE& l2, const lts_preorder pre, con
     {
       if (generate_counter_example)
       {
-        detail::counter_example_constructor cec("counter_example_failures_refinement.trc");
+        detail::counter_example_constructor cec(counter_example_file_name != "" ? counter_example_file_name : "counter_example_failures_refinement.trc");
         return destructive_refinement_checker(l1, l2, failures, false, cec);
       }
       return destructive_refinement_checker(l1, l2, failures, false);
@@ -845,7 +847,7 @@ bool destructive_compare(LTS_TYPE& l1, LTS_TYPE& l2, const lts_preorder pre, con
     {
       if (generate_counter_example)
       {
-        detail::counter_example_constructor cec("counter_example_weak_failures_refinement.trc");
+        detail::counter_example_constructor cec(counter_example_file_name != "" ? counter_example_file_name : "counter_example_weak_failures_refinement.trc");
         return destructive_refinement_checker(l1, l2, failures, true, cec);
       }
       return destructive_refinement_checker(l1, l2, failures, true);
@@ -854,7 +856,7 @@ bool destructive_compare(LTS_TYPE& l1, LTS_TYPE& l2, const lts_preorder pre, con
     {
       if (generate_counter_example)
       {
-        detail::counter_example_constructor cec("counter_example_failures_divergence_refinement.trc");
+        detail::counter_example_constructor cec(counter_example_file_name != "" ? counter_example_file_name : "counter_example_failures_divergence_refinement.trc");
         return destructive_refinement_checker(l1, l2, failures_divergence, true,cec);
       }
       return destructive_refinement_checker(l1, l2, failures_divergence, true);
diff --git a/tools/release/ltscompare/ltscompare.cpp b/tools/release/ltscompare/ltscompare.cpp
index f819f05..6c454c8 100644
--- a/tools/release/ltscompare/ltscompare.cpp
+++ b/tools/release/ltscompare/ltscompare.cpp
@@ -57,6 +57,7 @@ struct t_tool_options
   lts_preorder    preorder;
   std::vector<std::string> tau_actions;   // Actions with these labels must be considered equal to tau.
   bool generate_counter_examples;
+  std::string counter_example_file_name;
 };
 
 typedef  input_tool ltscompare_base;
@@ -133,7 +134,7 @@ class ltscompare_tool : public ltscompare_base
         mCRL2log(verbose) << "comparing LTSs using " <<
                      description(tool_options.preorder) << "..." << std::endl;
 
-        result = compare(l1,l2,tool_options.preorder,tool_options.generate_counter_examples);
+        result = compare(l1,l2,tool_options.preorder,tool_options.generate_counter_examples, tool_options.counter_example_file_name);
 
         mCRL2log(info) << "The LTS in " << tool_options.name_for_first
                        << " is " << ((result) ? "" : "not ")
@@ -257,7 +258,9 @@ class ltscompare_tool : public ltscompare_base
                  "be internal (tau) actions in addition to those defined as such by "
                  "the input").
       add_option("counter-example",
-                 "generate counter example traces if the input lts's are not equivalent",'c');
+                 "generate counter example traces if the input lts's are not equivalent",'c')
+      .add_option("out-counter-example", make_mandatory_argument("FILE-NAME"),
+                  "save counter exaple to FILE-NAME.");
     }
 
     void parse_options(const command_line_parser& parser)
@@ -357,6 +360,7 @@ class ltscompare_tool : public ltscompare_base
       {
         tool_options.format_for_second = mcrl2::lts::detail::guess_format(tool_options.name_for_second);
       }
+      tool_options.counter_example_file_name = parser.option_argument("out-counter-example");
     }
 
 };
-- 
Jan Nieuwenhuizen <janneke@gnu.org> | GNU LilyPond http://lilypond.org
Freelance IT http://JoyofSource.com | Avatar® http://AvatarAcademy.com

